File: decision-matrix/backend/app/geo/coord_equal.py
"""Exact coordinate equality for line ↔ point attachment (no distance tolerance)."""

from __future__ import annotations

import math

# ~1 mm at equator in degrees
COORD_EQUAL_ABS_TOL = 1e-8


def coords_equal(lon1: float, lat1: float, lon2: float, lat2: float) -> bool:
    """True if two WGS84 points are the same stored position."""
    return math.isclose(lon1, lon2, abs_tol=COORD_EQUAL_ABS_TOL, rel_tol=0.0) and math.isclose(
        lat1, lat2, abs_tol=COORD_EQUAL_ABS_TOL, rel_tol=0.0
    )

File: decision-matrix/frontend/src/lib/coordEqual.ts
import { linkCoordMatch } from './infraLinks';

/** ~1 mm at equator in degrees — matches backend coord_equal.py */
export const COORD_EQUAL_ABS_TOL = 1e-8;

export function coordsEqual(
  lon1: number,
  lat1: number,
  lon2: number,
  lat2: number,
): boolean {
  return (
    Math.abs(lon1 - lon2) <= COORD_EQUAL_ABS_TOL &&
    Math.abs(lat1 - lat2) <= COORD_EQUAL_ABS_TOL
  );
}

/** Exact match or same rounded display (legacy lines saved with 3-decimal ends). */
export function coordsEqualOrRounded(
  lon1: number,
  lat1: number,
  lon2: number,
  lat2: number,
): boolean {
  return (
    coordsEqual(lon1, lat1, lon2, lat2) ||
    (linkCoordMatch(lon1, lon2) && linkCoordMatch(lat1, lat2))
  );
}
